librz/il/vm_layer.c

```diff
@@ -0,0 +1,484 @@
+	ops_options.freefn = free_opcode_kv;
+	ops_options.elem_size = sizeof(HtPPKv);
+	vm->ct_opcodes = ht_pp_new_opt(&ops_options);
+
+	// init jump table of labels
+	vm->op_handler_table = (RzILOpHandler *)malloc(sizeof(RzILOpHandler) * RZIL_OP_INVALID);
+	memset(vm->op_handler_table, 0, RZIL_OP_INVALID);
+	vm->op_handler_table[RZIL_OP_VAR] = &rz_il_handler_var;
+	vm->op_handler_table[RZIL_OP_ITE] = &rz_il_handler_ite;
+	vm->op_handler_table[RZIL_OP_UNK] = &rz_il_handler_unk;
+
+	vm->op_handler_table[RZIL_OP_B0] = &rz_il_handler_b0;
+	vm->op_handler_table[RZIL_OP_B1] = &rz_il_handler_b1;
+	vm->op_handler_table[RZIL_OP_AND_] = &rz_il_handler_and_;
+	vm->op_handler_table[RZIL_OP_OR_] = &rz_il_handler_or_;
+	vm->op_handler_table[RZIL_OP_INV] = &rz_il_handler_inv;
+
+	vm->op_handler_table[RZIL_OP_LOAD] = &rz_il_handler_load;
+	vm->op_handler_table[RZIL_OP_STORE] = &rz_il_handler_store;
+
+	vm->op_handler_table[RZIL_OP_INT] = &rz_il_handler_int;
+	vm->op_handler_table[RZIL_OP_NEG] = &rz_il_handler_neg;
+	vm->op_handler_table[RZIL_OP_NOT] = &rz_il_handler_not;
+	vm->op_handler_table[RZIL_OP_LSB] = &rz_il_handler_lsb;
+	vm->op_handler_table[RZIL_OP_MSB] = &rz_il_handler_msb;
+	vm->op_handler_table[RZIL_OP_SHIFTL] = &rz_il_handler_shiftl;
+	vm->op_handler_table[RZIL_OP_SHIFTR] = &rz_il_handler_shiftr;
+	vm->op_handler_table[RZIL_OP_ADD] = &rz_il_handler_add;
+	vm->op_handler_table[RZIL_OP_SUB] = &rz_il_handler_sub;
+	vm->op_handler_table[RZIL_OP_MUL] = &rz_il_handler_mul;
+	vm->op_handler_table[RZIL_OP_DIV] = &rz_il_handler_div;
+	vm->op_handler_table[RZIL_OP_MOD] = &rz_il_handler_mod;
+	vm->op_handler_table[RZIL_OP_SDIV] = &rz_il_handler_sdiv;
+	vm->op_handler_table[RZIL_OP_SMOD] = &rz_il_handler_smod;
+
+	vm->op_handler_table[RZIL_OP_PERFORM] = &rz_il_handler_perform;
+	vm->op_handler_table[RZIL_OP_SET] = &rz_il_handler_set;
+	vm->op_handler_table[RZIL_OP_GOTO] = &rz_il_handler_goto;
+	vm->op_handler_table[RZIL_OP_BRANCH] = &rz_il_handler_branch;
+	vm->op_handler_table[RZIL_OP_SEQ] = &rz_il_handler_seq;
+
+	vm->var_count = 0;
+	vm->val_count = 0;
+	vm->mem_count = 0;
+	vm->easy_debug = 0;
+
+	return true;
+}
+
+/**
+ * Close and clean vm
+ * \param vm RzILVM pointer to VM
+ */
+RZ_API void rz_il_vm_close(RzILVM vm) {
+	RzILVar var;
+
+	rz_il_free_bag(vm->vm_global_value_set);
+
+	for (int i = 0; i < VM_MAX_VAR; ++i) {
+		if (vm->vm_global_variable_list[i] != NULL) {
+			var = vm->vm_global_variable_list[i];
+			rz_il_free_variable(var);
+			vm->vm_global_variable_list[i] = NULL;
+		}
+	}
+	free(vm->vm_global_variable_list);
+
+	if (vm->ct_opcodes) {
+		ht_pp_free(vm->ct_opcodes);
+	}
+
+	if (vm->mems) {
+		for (int i = 0; i < vm->mem_count; ++i) {
+			rz_il_free_mem(vm->mems[i]);
+		}
+		free(vm->mems);
+	}
+
+	if (vm->temp_value_list != NULL) {
+		for (int i = 0; i < VM_MAX_TEMP; ++i) {
+			free(vm->temp_value_list[i]);
+		}
+		free(vm->temp_value_list);
+	}
+
+	if (vm->vm_global_bind_table != NULL) {
```

Comment:
Remove these extra checks.

Reply:
@Heersin don't forget about these as well.

Reply:
@Heersin don't forget about this

Reply:
@Heersin please don't forget about this